diff --git a/pkg/registry/pod/rest.go b/pkg/registry/pod/rest.go
index c3c14b6..8138bae 100644
--- a/pkg/registry/pod/rest.go
+++ b/pkg/registry/pod/rest.go
@@ -234,7 +234,7 @@ func (rs *REST) fillPodInfo(pod *api.Pod) {
 		if ok {
 			if netContainerInfo.PodIP != "" {
 				pod.Status.PodIP = netContainerInfo.PodIP
-			} else {
+			} else if netContainerInfo.State.Running != nil {
 				glog.Warningf("No network settings: %#v", netContainerInfo)
 			}
 		} else {
