diff --git a/pkg/kubelet/util.go b/pkg/kubelet/util.go
index cc6aa58..0fa4fbe 100644
--- a/pkg/kubelet/util.go
+++ b/pkg/kubelet/util.go
@@ -98,14 +96,17 @@ func SetupLogging() {
 }
 
 // TODO: move this into pkg/client
-func getApiserverClient(authPath string, apiServerList util.StringList) (*client.Client, error) {
-	authInfo, err := clientauth.LoadFromFile(authPath)
-	if err != nil {
-		return nil, err
-	}
-	clientConfig, err := authInfo.MergeWithConfig(client.Config{})
-	if err != nil {
-		return nil, err
+func GetApiserverClient(authPath string, apiServerList util.StringList) (*client.Client, error) {
+	clientConfig := client.Config{}
+	if authPath != "" {
+		authInfo, err := clientauth.LoadFromFile(authPath)
+		if err != nil {
+			return nil, err
+		}
+		clientConfig, err = authInfo.MergeWithConfig(clientConfig)
+		if err != nil {
+			return nil, err
+		}
 	}
 	if len(apiServerList) < 1 {
 		return nil, fmt.Errorf("no apiservers specified.")
@@ -127,7 +128,7 @@ func SetupEventSending(authPath string, apiServerList util.StringList) {
 	if len(apiServerList) < 1 {
 		glog.Info("No api servers specified.")
 	} else {
-		if apiClient, err := getApiserverClient(authPath, apiServerList); err != nil {
+		if apiClient, err := GetApiserverClient(authPath, apiServerList); err != nil {
 			glog.Errorf("Unable to make apiserver client: %v", err)
 		} else {
 			// Send events to APIserver if there is a client.
