diff --git a/pkg/cloudprovider/controller/nodecontroller.go b/pkg/cloudprovider/controller/nodecontroller.go
index 2b8dec2..dd242fe 100644
--- a/pkg/cloudprovider/controller/nodecontroller.go
+++ b/pkg/cloudprovider/controller/nodecontroller.go
@@ -340,7 +340,7 @@ func (s *NodeController) deletePods(nodeID string) error {
 		}
 		glog.V(2).Infof("Delete pod %v", pod.Name)
 		if err := s.kubeClient.Pods(api.NamespaceAll).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v", pod.Name)
+			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
 		}
 	}
 
