diff --git a/pkg/hyperkube/server.go b/pkg/hyperkube/server.go
index 697211b..84c7e80 100644
--- a/pkg/hyperkube/server.go
+++ b/pkg/hyperkube/server.go
@@ -17,6 +17,7 @@ limitations under the License.
 package hyperkube
 
 import (
+	"errors"
 	"io/ioutil"
 	"strings"
 
@@ -71,3 +72,10 @@ func (s *Server) Flags() *pflag.FlagSet {
 	}
 	return s.flags
 }
+
+func (s *Server) FindServer(name string) (*Server, error) {
+	if s == nil {
+		return nil, errors.New("no hyperkube server present")
+	}
+	return s.hk.FindServer(name)
+}
