diff --git a/cmd/kube-apiserver/plugins.go b/cmd/kube-apiserver/plugins.go
index 64908e5..45ccfa4 100644
--- a/cmd/kube-apiserver/plugins.go
+++ b/cmd/kube-apiserver/plugins.go
@@ -25,4 +25,5 @@ import (
 	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/openstack"
 	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/ovirt"
 	_ "github.com/GoogleCloudPlatform/kubernetes/pkg/cloudprovider/vagrant"
+	_ "github.com/mesosphere/kubernetes-mesos/pkg/cloud/mesos"
 )
