diff --git a/pkg/cloudprovider/controller/nodecontroller.go b/pkg/cloudprovider/controller/nodecontroller.go
index 2b8dec2..c22e8fa 100644
--- a/pkg/cloudprovider/controller/nodecontroller.go
+++ b/pkg/cloudprovider/controller/nodecontroller.go
@@ -339,8 +339,8 @@ func (s *NodeController) deletePods(nodeID string) error {
 			continue
 		}
 		glog.V(2).Infof("Delete pod %v", pod.Name)
-		if err := s.kubeClient.Pods(api.NamespaceAll).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v", pod.Name)
+		if err := s.kubeClient.Pods(pod.Namespace).Delete(pod.Name); err != nil {
+			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
 		}
 	}
 
