diff --git a/messenger/messenger.go b/messenger/messenger.go
index 8cd5b27..805f895 100644
--- a/messenger/messenger.go
+++ b/messenger/messenger.go
@@ -234,7 +234,9 @@ func (m *MesosMessenger) reportError(err error) {
 	case <-ctx.Done():
 		<-c // wait for Route to return
 	case e := <-c:
-		log.Errorf("failed to report error %v due to: %v", err, e)
+		if e != nil {
+			log.Errorf("failed to report error %v due to: %v", err, e)
+		}
 	}
 }
 
