diff --git a/messenger/messenger.go b/messenger/messenger.go
index caf2af9..8cd5b27 100644
--- a/messenger/messenger.go
+++ b/messenger/messenger.go
@@ -234,7 +234,7 @@ func (m *MesosMessenger) reportError(err error) {
 	case <-ctx.Done():
 		<-c // wait for Route to return
 	case e := <-c:
-		log.Error("failed to report error %v due to: %v", err, e)
+		log.Errorf("failed to report error %v due to: %v", err, e)
 	}
 }
 
diff --git a/scheduler/scheduler.go b/scheduler/scheduler.go
index 44878b0..cdefc6d 100644
--- a/scheduler/scheduler.go
+++ b/scheduler/scheduler.go
@@ -662,7 +662,10 @@ func (driver *MesosSchedulerDriver) KillTask(taskId *mesos.TaskID) (mesos.Status
 		return driver.Status(), fmt.Errorf("Not connected to master")
 	}
 
-	message := &mesos.KillTaskMessage{TaskId: taskId}
+	message := &mesos.KillTaskMessage{
+		FrameworkId: driver.FrameworkInfo.Id,
+		TaskId:      taskId,
+	}
 
 	if err := driver.send(driver.MasterPid, message); err != nil {
 		log.Errorf("Failed to send KillTask message: %v\n", err)
